### DISCUSSION TITLE
How to Register a New JSON-RPC Method (API) Using a Plugin

### DISCUSSION THREAD

**Discussion by e-labInnovations** [kanboard/kanboard#5525] [Q&A] [answered]
Hello everyone,
I am looking to extend Kanboard's functionality by adding a new `JSON-RPC method` through a plugin. I've gone through the Kanboard documentation but couldn't find detailed instructions on how to register a custom JSON-RPC method. 

Can anyone guide me on how to:
1. Create a new JSON-RPC method.
2. Register this method within a Kanboard plugin.
3. Ensure the new method is accessible through the JSON-RPC endpoint.

Any code examples or links to relevant documentation would be greatly appreciated. 

Thank you!

**Comment by fguillot** [comment, marked as answer]
Example to add a new JSON-RPC method `myAPIProcedure` with a plugin:

```php
<?php

namespace Kanboard\Plugin\MyPlugin;

use Kanboard\Core\Plugin\Base;

class Plugin extends Base
{
    public function initialize()
    {
        $this->api->getProcedureHandler()->withCallback('myAPIProcedure', function($a, $b) {
            return 'a + b = '.($a + $b);
        });
    }
}
```

Example to call the new API procedure:

```bash
curl -s \
  -u admin:admin \
  -d '{"jsonrpc": "2.0", "method": "myAPIProcedure", "id": 1, "params": {"a": 2, "b": 3}}' \
  http://example.org/jsonrpc.php
{
  "jsonrpc": "2.0",
  "result": "a + b = 5",
  "id": 1
}
```

For reference:
- https://github.com/kanboard/kanboard/tree/main/libs/jsonrpc
- https://github.com/kanboard/kanboard/blob/main/app/ServiceProvider/ApiProvider.php
